File: data-structure-and-algorithm/cpp/largest-rectangle-in-histogram.cpp
// 84. Largest Rectangle in Histogram
// Difficulty: Hard

// Given n non-negative integers representing the histogram bar height where the width of each bar is 1,
// find the area of largest rectangle in the histogram.

// For example,
// Given heights = [1,3,2,4],
// return 6.

// |
// |            __
// |      __   |  |
// |     |  |__|  |
// |   __|  |  |  |
// |__|__|__|__|__|___


// Time:  O(n)
// Space: O(n)

class Solution {
public:
  int largestRectangleArea(vector<int>& heights) {
    heights.push_back(0); // for final calculate
    stack<int> idx;
    int maxArea = 0;

    for (int i = 0; i < heights.size();) {
      if (idx.empty() || heights[i] > heights[idx.top()]) {
        idx.emplace(i++);
      } else {
        int height = heights[idx.top()]; idx.pop();
        int len = idx.empty() ? i : i - idx.top() - 1;
        maxArea = max(maxArea, height * len);
      }
    }
    return maxArea;
  }
};

File: data-structure-and-algorithm/java/largest-rectangle-in-histogram.java
/**
 * Largest Rectangle in Histogram
 * 
 * Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.
 */

 import java.util.Stack;

public class Solution {
    public int largestRectangleArea(int[] height) {
        if (height == null) {
            throw new NullPointerException();
        }

        // add an zero element at the back to conviently calculate for the last height
        int[] heightList = new int[height.length + 1];
        for (int i = 0; i < height.length; ++i) {
            heightList[i] = height[i];
        }

        int maxArea = 0;
        Stack<Integer> heightIndexStack = new Stack<>();

        for (int i = 0; i < heightList.length; ++i) {
            while(!heightIndexStack.isEmpty() && heightList[heightIndexStack.peek()] >= heightList[i]) {
                int curr = heightIndexStack.pop();
                maxArea = Math.max(maxArea, heightList[curr] * (heightIndexStack.isEmpty() ? i : i - heightIndexStack.peek() - 1));
            }
            heightIndexStack.push(i);
        }

        return maxArea;
    }
}
